### PR TITLE
add pre-copy phase in live migration example

### PR DESCRIPTION
**Pull request by tmakatos**
This patch adds a simplistic pre-copy phase in the live migration
sample. The end goal is have a separate thread in the client to
modify device state while the device is in the pre-copy phase. This will
be done in later patches.

BAR1 is saved during the pre-copy phase and BAR0 is saved during the
stop-and-copy phase. This is purely for convenience. There are quite a
few assumptions and FIXMEs, even in the client code. We plan to address
them in future patches.

Signed-off-by: Thanos Makatos <thanos.makatos@nutanix.com>

### CODE REVIEW

File: samples/client.c [hunk 1]
```diff
@@ -716,20 +716,29 @@ usage(char *path) {
             basename(path));
 }
 
-static void
-migrate_from(int sock, int migr_reg_index, void **data, __u64 *len)
+/*
+ * Normally each time the source client (QEMU) would read migration data from
+ * the device it would send them to the destination cliet. However, since in
```
Comment (jlevon) [inline samples/client.c:721]: nit, "client"

File: samples/client.c [hunk 2]
```diff
@@ -716,20 +716,29 @@ usage(char *path) {
...
+ * our sample both the source and the destination client are the same process,
+ * we simply accumulate the migration data of each iteration and apply it to
+ * the destination server at the end.
+ *
+ * Performs as many migration loops as @nr_iters or until the device has no
+ * more migration data (pending_bytes is zero), which ever comes first. The
+ * result of each migration iteration is stored in @migr_iter.  @migr_iter must
+ * be at least of @nr_iters.
```
Comment (jlevon) [inline samples/client.c:729]: nit, s/ of// ?

File: samples/client.c [hunk 3]
```diff
@@ -799,28 +802,96 @@ migrate_from(int sock, int migr_reg_index, void **data, __u64 *len)
...
+                        offsetof(struct vfio_device_migration_info, device_state),
+                        &device_state, sizeof(device_state));
+    if (ret < 0) {
+        errx(EXIT_FAILURE, "failed to write to device state: %s",
+             strerror(-ret));
+    }
+
+    _nr_iters += do_migrate(sock, migr_reg_index, 1, (*migr_iters) + _nr_iters);
+
+    /* XXX read device state, migration must have finished now */
```
Comment (jlevon) [inline samples/client.c:871]: Can't we check _nr_iters again

File: samples/server.c [hunk 4]
```diff
@@ -299,20 +319,47 @@ static size_t
...
-    if (offset == 0 && size >= sizeof server_data->bar0) {
-        memcpy(buf, &server_data->bar0, sizeof server_data->bar0);
+    if (server_data->migration.state == VFU_MIGR_STATE_PRE_COPY) {
+        p = server_data->bar1;
+        bar_size = server_data->bar1_size;
+    } else if (server_data->migration.state == VFU_MIGR_STATE_STOP_AND_COPY) {
+        p = (uint8_t*)&server_data->bar0;
+        bar_size = sizeof server_data->bar0;
+    } else {
+        /*
```
Comment (jlevon) [inline samples/server.c:349]: add a log entry?

File: samples/server.c [hunk 5]
```diff
@@ -324,41 +371,44 @@ migration_write_data(vfu_ctx_t *vfu_ctx, void *data, __u64 size, __u64 offset)
...
-                          sizeof server_data->bar0, 0, true);
-        if (ret < 0) {
-            return ret;
-        }
-    }
+    /*
+     * We apply migration state directly in the migration_write_data callback,
+     * we don't need to do anything here. We would have to apply migration
+     * state in this callback if the migration region was memory mappable, in
+     * which we wouldn't know when the client wrote migration data.
```
Comment (jlevon) [inline samples/server.c:410]: "in which" -> "in which case"

File: samples/server.c [hunk 6]
```diff
@@ -369,10 +419,11 @@ int main(int argc, char *argv[])
     bool verbose = false;
     char opt;
     struct sigaction act = {.sa_handler = _sa_handler};
+    size_t bar1_size = 0x3000;
     struct server_data server_data = {
         .migration = {
             /* one page so that we can memory map it */
```
Comment (jlevon) [inline samples/server.c:425]: stale comment

File: samples/server.c [hunk 7]
```diff
@@ -431,21 +476,29 @@ int main(int argc, char *argv[])
 
     /*
      * Setup BAR1 to be 3 pages in size where only the first and the last pages
-     * are mappable.
+     * are mappable. The client can still mmap the 2nd page, we can't prohibit
+     * this under Linux. If we really want to probihit it we have to use
```
Comment (jlevon) [inline samples/server.c:480]: nit, "prohibit"

File: samples/server.c [hunk 8]
```diff
@@ -431,21 +476,29 @@ int main(int argc, char *argv[])
...
+     * separate files for the same region.
      */
     if ((fp = tmpfile()) == NULL) {
         err(EXIT_FAILURE, "failed to create BAR1 file");
     }
-    if (ftruncate(fileno(fp), 0x3000) == -1) {
+    server_data.bar1_size = 0x3000;
```
Comment (jlevon) [inline samples/server.c:486]: use "bar1_size"?

**Review by jlevon — CHANGES_REQUESTED** [review]
Just a few nits!

File: samples/client.c (hunk 1 above)
Comment (tmakatos) [reply samples/client.c:721]: ok

File: samples/client.c (hunk 2 above)
Comment (tmakatos) [reply samples/client.c:729]: ok

File: samples/server.c (hunk 4 above)
Comment (tmakatos) [reply samples/server.c:349]: We shouldn't call the device callaback in the first place, I'll add a FIXME in the library code.

File: samples/server.c (hunk 5 above)
Comment (tmakatos) [reply samples/server.c:410]: ok

File: samples/server.c (hunk 6 above)
Comment (tmakatos) [reply samples/server.c:425]: ok

File: samples/server.c (hunk 7 above)
Comment (tmakatos) [reply samples/server.c:480]: ok

File: samples/server.c (hunk 8 above)
Comment (tmakatos) [reply samples/server.c:486]: ok

File: samples/client.c (hunk 3 above)
Comment (tmakatos) [reply samples/client.c:876]: sure

**Pull request merged**